File: LRU CACHE/LRUCache.java
import java.util.*;
class Node{
    int key;
    int value;
    Node next;
    Node prev;
    public Node(int key,int value){
        this.key = key;
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}
class LRUCache {
    int capacity;
    HashMap<Integer,Node> cache;
    Node head;
    Node tail;
    public LRUCache(int capacity) {
        cache = new HashMap<>();
        this.capacity = capacity;

        head = new Node(0,0);
        tail = new Node(0,0);

        head.next = tail;
        tail.prev = head;
    }
    public void insert(Node node){
        Node prevLast = tail.prev;

        prevLast.next = node;
        node.prev = prevLast;

        node.next = tail;
        tail.prev = node;
    }
    public void remove(Node node){
        Node prevNode = node.prev;
        Node nextNode = node.next;

        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }

    public int get(int key) {
        if(!cache.containsKey(key)){
            return -1;
        }
        Node node = cache.get(key);
        remove(node);
        insert(node);
        return node.value;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)){
            Node oldNode = cache.get(key);
            remove(oldNode);
        }
        Node node = new Node(key,value);
        cache.put(key,node);
        insert(node);
        if(cache.size() > capacity){
            Node lru = head.next;
            head.next = lru.next;
            remove(lru);
            cache.remove(lru.key);
        } 

    }
}

// Your code will be instantiated and invoked as follows:
// LRUCache cache = new LRUCache(2);
// cache.put(1, 1);
// cache.put(2, 2);
// cache.get(1);       // returns 1
// cache.put(3, 3);    // evicts key 2
// cache.get(2);       // returns -1

File: LRU CACHE/LRUCache.py
class Node:
    def __init__(self,key,value):
            self.prev = None
            self.next = None
            self.key = key
            self.value = value
class LRUCache:
    def __init__(self, capacity: int):
        self.cache = {}
        self.capacity = capacity

        self.head = Node(0,0)
        self.tail = Node(0,0)

        self.head.next = self.tail
        self.tail.prev = self.head
    def insert(self,node):
        prevLast = self.tail.prev

        prevLast.next = node
        node.prev = prevLast
        
        node.next = self.tail
        self.tail.prev = node
    def remove(self,node):
        prevNode = node.prev
        nextNode = node.next

        prevNode.next = nextNode
        nextNode.prev = prevNode
    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        
        node = self.cache[key]
        self.remove(node)
        self.insert(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            oldNode = self.cache[key]
            self.remove(oldNode)
        node = Node(key,value)
        self.cache[key] = node
        self.insert(node)
        if len(self.cache) > self.capacity:
            lru = self.head.next
            self.remove(lru)
            del self.cache[lru.key]

## Your code will be instantiated and invoked as follows:
# cache = LRUCache(2)
# cache.put(1, 1)
# cache.put(2, 2)
# cache.get(1)       # returns 1
# cache.put(3, 3)    # evicts key 2
# cache.get(2)       # returns -1
